### PR TITLE
[Code cleaning]: Replace "raw" pointers by using "pointer" library

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
This PR replaces raw pointers with pointers that are created by using the [pointer](https://pkg.go.dev/k8s.io/utils/pointer) library which is part of the [k8s utils repo](https://github.com/kubernetes/utils).

In general, the changes look something like this:
Before this PR:
```go
_true := true
s := SomeStruct{
    SomeBoolPtrField: &_true
}
```

After this PR:
```go
s := SomeStruct{
    SomeBoolPtrField: pointer.BoolPtr(true)
}
```

The changes bring two main advantages:
1. Nicer code. Using things like `_true := true` is ugly, unclear, and adds noise to the code.
2. With `pointer` library the values are cached, so only one pointer is made for `false` and one for `true`. This also applies to integers, strings, etc. Many values are re-used all the time (e.g. `0`, `""`, etc) and this can grant a performance boost.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
I must admit, I do not like such dependencies on external libraries.
If you find it really better, I would just clone it in the project over depending on it.

I saw we already depend on it in many places, so this is not something new, however, the fact that we are over-depending on external packages & modules does not mean we should continue doing so.

My reasoning includes:
- Too many dependencies on external tooling is exposing us to changes that may require from us more work in the future (e.g. deprecation, support for new golang version only, etc).
For cases like these, I would avoid it or just copy it in the project.
- It is hard to preserve consistency and enforce using this package and not the direct approach.
- I am not convinced it is cleaner or faster as this change claims, or at the least that it has any affect on the system in terms of efficiency.

Also, it is worth noting a deprecation that was missed: All `*Ptr` functions got replaced by shorter versions (e.g. `Bool()` replacing `BoolPtr`). This is a classic example where one would prefer not to be dependent on external libraries that are very simple to maintain in-house.
(although not removed, it is reasonable that it will happen sometime in the future)

**Comment by iholder-redhat** [comment]
> I must admit, I do not like such dependencies on external libraries. If you find it really better, I would just clone it in the project over depending on it.
> 
> I saw we already depend on it in many places, so this is not something new, however, the fact that we are over-depending on external packages & modules does not mean we should continue doing so.
> 
> My reasoning includes:
> 
> * Too many dependencies on external tooling is exposing us to changes that may require from us more work in the future (e.g. deprecation, support for new golang version only, etc).
>   For cases like these, I would avoid it or just copy it in the project.
> * It is hard to preserve consistency and enforce using this package and not the direct approach.
> * I am not convinced it is cleaner or faster as this change claims, or at the least that it has any affect on the system in terms of efficiency.
> 
> Also, it is worth noting a deprecation that was missed: All `*Ptr` functions got replaced by shorter versions (e.g. `Bool()` replacing `BoolPtr`). This is a classic example where one would prefer not to be dependent on external libraries that are very simple to maintain in-house. (although not removed, it is reasonable that it will happen sometime in the future)

I feel the discussion you're raising is more general. So here's my general opinion:
I think your concern is valid and has a lot of truth to it, but I feel you only show one side of the trade-off here.

You're right IMO - using too many dependencies can cause damage and has its downsides, like keeping track of another project (when to update, what to deprecate, etc) and of course we trust someone else to do something for us. I also agree that the performance boost is theoretical - until we start profiling we can't know if in reality things are actually faster.

But of course, there's another side to this trade-off, because we can't (and shouldn't) do everything ourselves. The principle of cohesiveness (which is also the unix' motto: "make each program do one thing well") means that except for your main business logic (the "one thing" we do well) we **have** to trust others to perform tasks for us.

A classic example is string manipulations. Should we implement a package that manipulates strings for us? of course we could. And what about heavy mathematical calculations? Or security code that would integrate with the host OS to configure cgroups? Or code to handle http connections? Or to marshall / unmarshall objects?

I think that in general the rule of thumb should be something like: "Is it complected? Is it outside of our main business logic?" if the answer to both of these questions is "yes" then we probably shouldn't implement it. Caching of pointer values is not that different, in principle, then using k8s' informers and stores which also reside in an external library - should we implement informers as well? 

Bear in mind that like informers and stores the pointer library is an official k8s library, it's not that it's a non-maintained github repo that no one knows about.

By the way,
> I am not convinced it is cleaner or faster as this change claims

I can agree about the faster part (just because we didn't measure anything, although it's very probable it is faster and reduces memory footprint), but I don't agree at all that it isn't cleaner. I truly think `_true := true` is horrible, especially when it's spread around a thousand different places in our codebase.

**Comment by EdDev** [comment]
There is a balance that is suppose to be kept.
It is indeed beneficial not to re-invent the wheel each time, however, for simple syntactic sugar (which I believe this dependency is all about) it is not worth it IMO. You took the discussion to a generic direction, which is fine. But we are talking about this specific change which in my view is an overkill and introduces more maintenance burden then clarity.

I think case specifically, I think the ugliness comes from the language that does not provide this in a nice form. It is one of the cases where I think wrapping is not a good thing.
However, I recognize this is controversial and one will still prefer the syntactic sugar for saving one line of code, and for those I suggest going with the next best thing: using a self maintained `pointer` package over adding a dependency.

The disadvantages exist, you have even encountered it in this change (using deprecated names), so I would strongly suggest not to follow this path for this specific incident.

**Comment by iholder-redhat** [comment]
BTW,
> I this case specifically, I think the ugliness comes from the language that does not provide this in a nice form. It is one of the cases where I think wrapping is not a good thing.

Many programming languages official features are basically re-implementations of logic that previously existed in external libraries. This is very common and already happened in Go with the errors package that became official after many unofficial packages were out there. So I have to disagree that we have to surrender to a languages' ugly parts just because there is no current solution. Again, this library is part of k8s organization and very well known. It very well might be that it'll become an official language feature in the future.

But as I said, we're drifting away from this PR's scope. This conversation should be a community conversation about this general topic.

**Comment by xpivarc** [comment]
/hold
Please use the functions that are supported.

**Comment by iholder-redhat** [comment]
> /hold Please use the functions that are supported.

Thanks for the reminder!
/hold cancel

File: pkg/virt-api/rest/subresource_test.go [hunk 1]
```diff
@@ -93,9 +95,7 @@ var _ = Describe("VirtualMachineInstance Subresources", func() {
 	var vmiClient *kubecli.MockVirtualMachineInstanceInterface
 	var migrateClient *kubecli.MockVirtualMachineInstanceMigrationInterface
 
-	running := Running
-	notRunning := NotRunning
-	gracePeriodZero := int64(0)
+	gracePeriodZero := pointer.Int64Ptr(0)
```
Comment (xpivarc) [inline pkg/virt-api/rest/subresource_test.go:98]: `Int64Ptr` -> `Int64`

**Comment by xpivarc** [comment]
/hold
One more, pretty please!

**Comment by iholder-redhat** [comment]
> /hold One more, pretty please!

Done :smile: Thanks!
I'll let you unhold this time ;)

**Comment by xpivarc** [comment]
/hold cancel
Thanks

**Review by xpivarc — APPROVED** [review]

**Comment by dhiller** [comment]
sriov test has been successful once on the same commit: https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/7622/pull-kubevirt-e2e-kind-1.22-sriov/1518600155903823872

/override pull-kubevirt-e2e-kind-1.22-sriov

**Pull request merged**